### ISSUE TITLE
Should retry on errors.

### ISSUE THREAD

**Issue by teebu** [acupofjose/elasticstore#35]
I'm seeing this error, not sure why, but it should just keep retrying, at least a few times, instead of ending the process.

```
Error: Error 4: The datastore operation timed out, or the data was temporarily unavailable.
```

**Comment by acupofjose** [comment]
@teebu, I haven't seen the stream API for firestore queries before, so that's a cool find. Unfortunately, it looks like it is just for one-off queries to firestore, not for listeners. Elasticstore requires maintaining listeners for added/removed/modified. 

It looks like that has been updated upstream in `nodejs-firestore`, but I'm not sure it's made it into `firebase-admin` yet.

**Comment by teebu** [comment]
So you're saying the `nodejs-firestore` has the functionality of stream queue with onSnapshot listener to handling both streaming the query and listening for events? 

What about the solution provided in the ticket, https://github.com/googleapis/nodejs-firestore/issues/1040#issuecomment-618593346?

**Comment by acupofjose** [comment]
No, it doesn't have that functionality - or at least, doesn't have that functionality right now. Nor, unfortunately will the solution he proposes work for our purposes.

> "I need to not only get the first snapshot, but also subsequent updates, which makes the implementation more complex than the example you have given."
https://github.com/googleapis/nodejs-firestore/issues/1040#issuecomment-618603485

Now, an alternative might be to `stream` the initial data filling and then do an `onSnapshot` for data deltas only - but `onSnapshot` doesn't allow you to only listen for changes as [it is currently implemented](https://googleapis.dev/nodejs/firestore/latest/Query.html#onSnapshot).

That said, doing an `onSnapshot` on a large collection is supposed to be stable.

**Comment by teebu** [comment]
Some ideas from https://stackoverflow.com/questions/33885059/how-to-only-get-new-data-without-existing-data-from-a-firebase

One proposal is to have a `created_at` field in the collection, and you can use the where selector to now, or something like child_added, used here. https://github.com/FirebaseExtended/firechat/blob/master/src/js/firechat.js#L347

**Comment by thelaughingman** [comment]
Same problem here 😥
Any news on solutions/workarounds?

**Comment by teebu** [comment]
I contacted firebase support, their response: 

```Engineering team has suggested using limits and pagination to avoid this kind of errors. In these documents you will find more details about it.```

**Comment by acupofjose** [comment]
Well. Given that from support - any suggestions on how to make this work?

We could restart the listener in the event that it fails - but that would pull all the initial documents again. Which, in my mind, would have a lot of unnecessary reads. The `created_at`/`updated_at` field query is what I've implemented in the past to limit the number of reads and achieve long running listeners.

Frankly, I don't have a project that I'm able to see this error on. My example project https://elasticstore.netlify.app has long running listeners and has not had this problem.

Is there any way y'all can provide a project I can use to debug with?

**Comment by teebu** [comment]
Restarting the listener is a bad idea. I ran into an issue where I had this running on a kubernetes pod, and it ran out of memory relatively quickly, ended up restarting like 200 times in a couple of days, ended up costing like $100 bucks in firebase fees.

**Comment by acupofjose** [comment]
Ooof. Sorry to hear it. I'll keep mulling on it then. Thanks @teebu

**Comment by thelaughingman** [comment]
Thanks for the feedback guys. 
My collection in Firestore is relatively small, around 36k records, and the documents have 500/600 fields each, with seven sub-collections.

Is it too big/heavy?

@acupofjose if you wish I can give you access to my Firestore database

**Comment by timtim001hk** [comment]
Any update for this problem? 
I also encountered this error...In my case, the problem only occurs in collection with more than 100,000 records.

**Comment by timtim001hk** [comment]
Can this be solved in the following ways?

Firestore reactive pagination
Simple pagination for loading more content while still listening for updates
https://medium.com/@650egor/firestore-reactive-pagination-db3afb0bf42e

**Comment by acupofjose** [comment]
@timtim001hk - sorry for the delay, I've been out of the country. 

That's an interesting solution! I think it would require counting the number of records in the listener so that you can maintain the cursors. There is also a question of how big the listener blocks should be - 10,000 records? 50,000? 

Curious as to your thoughts on those!
